### ISSUE TITLE
CA2000 reported on disposable exceptions

### ISSUE THREAD

**Issue by lostmsu** [dotnet/roslyn-analyzers#3306]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.8 (Latest)

#### Diagnostic ID

Example: [CA2000](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2000)

#### Repro steps

```csharp
class DisposableException: Exception, IDisposable { public void Dispose(){} }
void Do(){ throw new DisposableException(); }
```

#### Expected behavior
No CA2000

#### Actual behavior
CA2000: call Dispose on object created by `new DisposableException()`

#### Notes
Exception leaves the scope.

**Comment by Evangelink** [comment]
I might say something stupid here but as exceptions denote an exceptional path which isn't likely to ever be disposed, what's the point of having the exception disposable?

**Comment by sharwell** [comment]
This is only a false positive if the caller fails to catch the exception and/or fails to dispose of the exception after it is caught. It's less clear whether the current behavior is desirable.

**Comment by lostmsu** [comment]
@sharwell I did not know FxCop performs analysis beyond the scope of a single function. For a public function in a library, not having any callers or having callers not catching exceptions is normal.

**Comment by lostmsu** [comment]
@Evangelink our use case is wrapping native exceptions. Our wrapper holds references to native resources of the original exception.

**Comment by Evangelink** [comment]
@lostmsu I see! I'm not trying to make you change anything but can't you simply close the resource when you have caught the native exception and translated everything you need/want into the CLR one?

**Comment by lostmsu** [comment]
@Evangelink the exception might itself link to a graph of objects of arbitrary size.
For a concrete example: a TensorFlow exception might link to the `Tensor`, that caused it. It might be a few gigabytes, so it would be infeasible to marshal it.

**Comment by Evangelink** [comment]
Thank you for sharing this use. It's not something I have ever done but that makes sense to me.

I tend to align with @sharwell on the fact that this is a pretty special case so I am not sure whether adding a built-in exclusion is the best solution.

For the time being, you can use the exclusion mechanism that was recently added (I am not sure whether this is already available in the latest public release), using for example `dotnet_code_quality.CA2000.excluded_symbol_names = DisposableException`
